kotlinx-coroutines-core/common/src/internal/SegmentList.kt

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+
+// returns the first segment `s` with `s.id >= id` or `CLOSED`
+// if all the segments in this linked list have lower `id` and the list is closed for further segment additions.
+private inline fun <S : Segment<S>> S.findSegmentInternal(id: Long, createNewSegment: (id: Long, prev: S?) -> S): SegmentOrClosed<S> {
+    // Go through `next` references and add new segments if needed,
+    // similarly to the `push` in the Michael-Scott queue algorithm.
+    // The only difference is that `CAS failure` means that the
+    // required segment has already been added, so the algorithm just
+    // uses it. This way, only one segment with each id can be added.
+    var cur: S = this
+    while (cur.id < id || cur.removed) {
+        val nextOrClosed = cur.nextOrClosed
+        if (nextOrClosed.isClosed) return SegmentOrClosed(CLOSED)
+        val next: S = if (nextOrClosed.segment === null) {
+            val newTail = createNewSegment(cur.id + 1, cur)
+            if (cur.trySetNext(newTail)) {
+                if (cur.removed) cur.remove()
+                newTail
+            } else {
+                val nextOrClosed = cur.nextOrClosed
+                if (nextOrClosed.isClosed) return SegmentOrClosed(CLOSED)
+                nextOrClosed.segment!!
+            }
+        } else nextOrClosed.segment!!
+        cur = next
+    }
+    return SegmentOrClosed(cur)
+}
+
+// Returns `false` if the segment `to` is logically removed, `true` on successful update.
+private inline fun <S : Segment<S>> AtomicRef<S>.moveForward(to: S): Boolean = loop { cur ->
+    if (cur.id >= to.id) return true
+    if (!to.tryIncPointers()) return false
+    if (compareAndSet(cur, to)) {
+        // the segment is moved
+        if (cur.decPointers()) cur.remove()
+        return true
+    } else {
+        if (to.decPointers()) to.remove()
```

Comment:
Assume the following execution:
* `cur` points to segment 0, and `to` has id 2 and is the physical tail. There are no pointers to `to` yet; for example, `cur` could be the tail pointer.
* `cur.moveForward(to)` is called. All the checks before the CAS pass.
* `cur.moveForward(segment_1)` is called and completed, where `segment_1` is the segment with `id = 1`.
* The CAS fails.
* The last cell is cancelled in the segment `to`.
* `to.decPointers()` returns `true`, and `to.remove()` is called, even though `to.removed` is not true.

This can never happen in the provided semaphore implementation because in order to cancel the cell, one must first leave the call to `suspend`, which moves the head pointer at least to the segment of the cell, and so the `head` points either to `to`, in which case `decPointer()` returns `false`, or to something to the right of it, in which case `to` is not a physical tail anymore and the problem isn't present.

Reply:
The problem occurs because `decPointers` has a bug -- it should not return true if the segment is the physical tail (see the documentation). Fixed this bug.